# Matrizes e vetores

## License

All content can be freely used and adapted under the terms of the 
[Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/).

![Creative Commons License](https://i.creativecommons.org/l/by/4.0/88x31.png)

## Representação de uma matriz

Antes de podermos manipular matrizes e vetores no computador, precisamos de um jeito de armazená-los em variáveis.
Para vetores, o candidato natural são **listas**.

$$
\mathbf{v} = \begin{bmatrix}1 \\ 2 \\ 3\end{bmatrix}
$$

O vetor acima poderia ser representado em código como:



In [1]:
v = [1, 2, 3]
print(v)

[1, 2, 3]


Uma matriz pode ser vista como um conjunto de vetores, ou um vetor de vetores. Cada vetor seria equivalente a uma linha da matriz:

$$
\mathbf{A} =
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9 \\
\end{bmatrix} =
\begin{bmatrix}
[1 & 2 & 3] \\
[4 & 5 & 6] \\
[7 & 8 & 9] \\
\end{bmatrix}
$$

Logo, um jeito de representar uma matriz em Python é através de uma **lista de listas**:

In [1]:
A = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print(A)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]


O Python permite quebrar a linha quando o comando está entre `[` ou `(`:

In [3]:
A = [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]
print(A)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]


*Note que o print acima imprime a matriz toda em uma única linha*.

A matriz `A` é uma lista como qualquer outra. A única diferença é que **os elementos da lista são outras listas**. O elemnto `[0]` da lista `A` será a lista correspondente a primeira linha:

In [4]:
print(A[0])

[1, 2, 3]


Como `A[0]` é uma lista, podemos pegar os elementos dessa lista da mesma forma:

In [5]:
print(A[0][0])
print(A[1][2])

1
6


### Exemplo

Queremos imprimir cada elemento da matriz `A` da seguinte forma:

    1  2  3  
    4  5  6  
    7  8  9

Podemos utilizar o `for` para pegar cada linha da matriz. Para cada linha da matriz, queremos imprimir o elemento. Quando terminarmos de imprimir a linha, vamos pular uma linha.

In [6]:
for i in range(3): # Anda sobre as linhas
    for j in range(3): # Anda sobre as colunas
        print(A[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() # Imprime nada e pula uma linha

1  2  3  
4  5  6  
7  8  9  


## Somando matrizes

A soma de duas matrizes produz uma matriz que terá a soma dos elementos correspondentes:

$$
\begin{bmatrix}
a & b & c \\
d & e & f \\
g & h & i \\
\end{bmatrix} +
\begin{bmatrix}
j & l & m \\
n & o & p \\
q & r & s \\
\end{bmatrix} =
\begin{bmatrix}
a + j & b + l & c + m \\
d + n & e + o & f + p \\
g + q & h + r & i + s \\
\end{bmatrix}
$$

De forma genérica, o $j$-ésimo elemento da $i$-ésima linha de uma matriz $\mathbf{A}$ é $A_{ij}$.
Se temos duas matrizes $\mathbf{A}$ e $\mathbf{B}$, a soma pode ser escrita como:

$$
C_{ij} = A_{ij} + B_{ij}
$$

## Tarefa

Some as matrizes `A` e `B` definidas abaixo e guarde o resultado em uma matriz `C`. Imprima a matriz `C`.

**Dicas**:

* Você pode criar a matriz `C` antes ou durante a soma. Lembre-se que a matriz é uma lista e listas possuem o método `append`.

In [2]:
A = [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]
B = [[3, 4, 5],
     [6, 7, 8],
     [9, 10, 11]]
nlin_a = 3
nlin_b = 3
ncol_a = 3
ncol_b = 3

In [3]:
C = [] #criamos a matriz C, que sera a resposta depois de preenchida

for i in range(3): #criamos um loop que repetira tres vezes os proximos comandos, passando pelas tres linhas da matriz A
    linha = [] #criamos uma linha vazia, que sera usada para preencher C
    for j in range(3): #criamos um novo loop dentro do anterior, que passara trez vezes, passando pelos tres elementos de cada linha.
        linha.append(A[i][j]+B[i][j]) #adicionamos na linha a soma de cada elemento de A e B
    C.append(linha) #apos terminarmos o loop e preenchermos a linha, adicionamos ela em C

print(C) #tcharam!

[[4, 6, 8], [10, 12, 14], [16, 18, 20]]


In [29]:
for i in range(3): # Anda sobre as linhas
    for j in range(3): # Anda sobre as colunas
        print(C[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() # Imprime nada e pula uma linha

4  6  8  
10  12  14  
16  18  20  


### Resultado esperado

Seu código deve imprimir exatamente:

    4  6  8  
    10  12  14  
    16  18  20  

## Multiplicando uma matriz por um vetor

A multiplicação de uma matriz por um vetor é

$$
\begin{bmatrix}
a & b  \\
c & d  \\
\end{bmatrix} 
\begin{bmatrix}
e \\
f \\
\end{bmatrix} =
\begin{bmatrix}
ae + bf \\
ce + df \\
\end{bmatrix}
$$

Sendo o vetor $\mathbf{u} = \mathbf{A}\mathbf{v}$, cada elemento $i$ de $\mathbf{u}$ é 


$$
u_i = \sum\limits_{k=1}^{N} A_{ik}v_k
$$

## Tarefa

Faça a multiplicação da matriz pelo vetor definidos abaixo. Guarde o resultado em uma lista.

**Dicas**

In [4]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
v = [12, 13, 14, 15]
nlin = 3
ncol = 4

In [1]:
C= [] #criamos a matriz resposta vazia
soma = 0 #criamos uma variavel vazia, usaremos ela dentro do loop

for i in range(nlin): #repetiremos i pelo numero de linhas da matriz A
    for j in range(ncol): #repetiremos j pelo numero de elementos em cada linha de A
        elemento = A[i][j]*v[j] #o elemento sera multiplicado pelo seu correspondente no vetor v
        soma = soma+elemento #para cada elemento multiplicado, somaremos na variavel soma
    C.append(soma) #a variavel final sera adicionada em C
    soma = 0 #zeramos soma para usar de novo no proximo loop

NameError: name 'nlin' is not defined

In [6]:
print(C) #tcharam!

[140, 302, 464]


### Resultado esperado

O seu código deve imprimir exatamente:

    [140, 302, 464]

## Multiplicação de matrizes

A multiplicação de matrizes é feita de forma diferente da soma. É mais fácil mostrar do que explicar:

$$
\begin{bmatrix}
a & b  \\
c & d  \\
\end{bmatrix} 
\begin{bmatrix}
e & f \\
g & h  \\
\end{bmatrix} =
\begin{bmatrix}
ae + bg & af + bh  \\
ce + dg & cf + dh  \\
\end{bmatrix}
$$

Sendo $\mathbf{C} = \mathbf{A}\mathbf{B}$, cada elemento $ij$ de $\mathbf{C}$ é 


$$
C_{ij} = \sum\limits_{k=1}^{N} A_{ik}B_{kj}
$$

## Tarefa

Faça a multiplicação das duas matrizes definidas abaixo. Guarde o resultado em uma matriz (lista de listas).

In [2]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
B = [[3, 4, 5],
     [6, 7, 8],
     [9, 10, 11],
     [12, 13, 14]]
nlin_a = 3
nlin_b = 4
ncol_a = 4
ncol_b = 3

In [3]:
C = [] #criamos C vazia
soma = 0 #criamos a variavel vazia

for i in range(3): #repetiremos o loop principal trez vezes, para cada linha de A
    linha = [] #criamos a linha vazia
    for j in range(3): #repetiremos o loop trez vezes, para cada elemento de cada linha de A
        for z in range(4): #repetiremos um novo loop quatro vezes, para cada coluna de B que sera multiplicada pelos elementos de A
            elemento = A[i][z]*B[z][j] #multiplicamos os valores desejados
            soma = soma+elemento #adicionamos na variavel soma ate que esteja completa
        linha.append(soma) #adicionamos a soma total na linha
        soma = 0 #zeramos a soma
    C.append(linha) # a linha, apos completa, sera adicionada em C
        
print(C) #tcharam!

[[90, 100, 110], [180, 202, 224], [270, 304, 338]]


In [4]:
for i in range(3): # Anda sobre as linhas
    for j in range(3): # Anda sobre as colunas
        print(C[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() # Imprime nada e pula uma linha

90  100  110  
180  202  224  
270  304  338  


### Resultado esperado

O seu código deve imprimir exatamente:

    90  100  110  
    180  202  224  
    270  304  338  



## Tarefa Bônus

Calcule o produto $\mathbf{A}^T\mathbf{A}$ para a matriz $A$ definida abaixo. $\mathbf{A}^T$ é a matriz transposta de $A$:

$$
\mathbf{A} = 
\begin{bmatrix}
a & b & c \\
d & e & f \\
\end{bmatrix}
$$

$$
\mathbf{A}^T = 
\begin{bmatrix}
a & d \\
b & e \\
c & f \\
\end{bmatrix}
$$

Ou seja, as linhas de $\mathbf{A}$ são as colunas de $\mathbf{A}^T$.


In [2]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
nlin = 3
ncol = 4

In [5]:
B = [[0,0,0],[0,0,0],[0,0,0],[0,0,0]] #criamos a transposta com todos os elementos em 0

for q in range(4): #criaremos a transposta a partir de A, passando por suas 4 linhas.
    for w in range(3): #passaremos por cada um dos tres elementos de cada linha de A.
        B[q][w] = A[w][q] #cada elemento da transposta B sera o correspondente de A
print(B) #testamos para ver se deu tudo certo...

C = [] #criamos a matriz C vazia
soma = 0 #criamos a variavel soma

for i in range(4): #repetiremos o loop principal quatro vezes, para cada linha de B
    linha = [] #criamos a linha vazia
    for j in range(4): #repetiremos o loop quatro vezes, para cada elemento de cada linha de B
        for z in range(3): #repetiremos um novo loop tres vezes, para cada coluna de A que sera multiplicada pelos elementos de B
            elemento = B[i][z]*A[z][j] #multiplicamos os valores desejados. Perceba que a transposta vem primeiro desssa vez
            soma = soma+elemento #adicionamos na variavel soma ate que esteja completa
        linha.append(soma) #adicionamos a soma total na linha
        soma = 0 #zeramos a soma
    C.append(linha) # a linha, apos completa, sera adicionada em C

[[1, 4, 7], [2, 5, 8], [3, 6, 9], [4, 7, 10]]


In [7]:
for i in range(4): # Anda sobre as linhas
    for j in range(4): # Anda sobre as colunas
        print(C[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() # Imprime nada e pula uma linha

66  78  90  102  
78  93  108  123  
90  108  126  144  
102  123  144  165  


### Resultado esperado

O seu código deve imprimir exatamente:

    66  78  90  102  
    78  93  108  123  
    90  108  126  144  
    102  123  144  165  
